reactos/drivers/usb/usbport/usbport.c

```diff
@@ -0,0 +1,2751 @@
+
+    FdoExtension = FdoDevice->DeviceExtension;
+
+    DmaAdapter = FdoExtension->DmaAdapter;
+    DmaOperations = DmaAdapter->DmaOperations;
+
+    HeaderSize = sizeof(USBPORT_COMMON_BUFFER_HEADER);
+    Length = ROUND_TO_PAGES(BufferLength + HeaderSize);
+    LengthPadded = Length - (BufferLength + HeaderSize);
+
+    BaseVA = (ULONG_PTR)DmaOperations->AllocateCommonBuffer(DmaAdapter,
+                                                            Length,
+                                                            &LogicalAddress,
+                                                            TRUE);
+
+    if (!BaseVA)
+        goto Exit;
+
+    StartBufferVA = BaseVA & ~(PAGE_SIZE - 1);
+    StartBufferPA = LogicalAddress.LowPart & ~(PAGE_SIZE - 1);
+
+    HeaderBuffer = (PUSBPORT_COMMON_BUFFER_HEADER)(StartBufferVA +
+                                                   BufferLength +
+                                                   LengthPadded);
+
+    HeaderBuffer->Length = Length;
+    HeaderBuffer->BaseVA = BaseVA;
+    HeaderBuffer->LogicalAddress = LogicalAddress;
+
+    HeaderBuffer->BufferLength = BufferLength + LengthPadded;
+    HeaderBuffer->VirtualAddress = StartBufferVA;
+    HeaderBuffer->PhysicalAddress = StartBufferPA;
+
+    RtlZeroMemory((PVOID)StartBufferVA, BufferLength + LengthPadded);
+
+Exit:
+    return HeaderBuffer;
+}
+
+VOID
+NTAPI
+USBPORT_FreeCommonBuffer(IN PDEVICE_OBJECT FdoDevice,
+                         IN PUSBPORT_COMMON_BUFFER_HEADER HeaderBuffer)
+{
+    PUSBPORT_DEVICE_EXTENSION FdoExtension;
+    PDMA_ADAPTER DmaAdapter;
+    PDMA_OPERATIONS DmaOperations;
+
+    DPRINT("USBPORT_FreeCommonBuffer: ... \n");
+
+    FdoExtension = FdoDevice->DeviceExtension;
+
+    DmaAdapter = FdoExtension->DmaAdapter;
+    DmaOperations = DmaAdapter->DmaOperations;
+
+    DmaOperations->FreeCommonBuffer(FdoExtension->DmaAdapter,
+                                    HeaderBuffer->Length,
+                                    HeaderBuffer->LogicalAddress,
+                                    (PVOID)HeaderBuffer->VirtualAddress,
+                                    TRUE);
+}
+
+PUSBPORT_MINIPORT_INTERFACE
+NTAPI
+USBPORT_FindMiniPort(IN PDRIVER_OBJECT DriverObject)
+{
+    KIRQL OldIrql;
+    PLIST_ENTRY List;
+    PUSBPORT_MINIPORT_INTERFACE MiniPortInterface = NULL;
+
+    DPRINT("USBPORT_FindMiniPort: ... \n");
+
+    KeAcquireSpinLock(&USBPORT_SpinLock, &OldIrql);
+
+    for (List = USBPORT_MiniPortDrivers.Flink;
+         List != &USBPORT_MiniPortDrivers;
+         List = List->Flink)
+    {
+        MiniPortInterface = CONTAINING_RECORD(List,
+                                              USBPORT_MINIPORT_INTERFACE,
+                                              DriverLink);
+
+        if (MiniPortInterface->DriverObject == DriverObject)
+        {
+            DPRINT("USBPORT_FindMiniPort: find MiniPortInterface - %p\n",
+                   MiniPortInterface);
+            break;
+        }
+    }
+
```

Comment:
This function will not return NULL if the miniport was not found. Needs a different variable for the result?